src/lib/handleRef.ts

```diff
@@ -0,0 +1,35 @@
+import * as React from 'react'
+
+/** A type that ensures that passed prop is defined in props and matches the type. */
+type HandleRefProps<TProp extends string, N> = { [K in TProp]?: React.Ref<N> }
+
+/**
+ * The function that correctly handles passing refs.
+ *
+ * @param props All components props
+ * @param propName The name of a ref prop
+ * @param node A node that should be passed by ref
+ */
+const handleRef = <P extends string, N>(props: HandleRefProps<P, N>, propName: P, node: N) => {
+  const ref: React.Ref<N> = props[propName]
+
+  if (process.env.NODE_ENV !== 'production') {
+    if (typeof ref === 'string') {
+      throw new Error(
+        'We do not support refs as string, this is a legacy API and will be likely to be removed in one of the future releases of React.',
+      )
+    }
+  }
+
+  if (typeof ref === 'function') {
+    ref(node)
+    return
+  }
+
+  if (typeof ref === 'object') {
+    // @ts-ignore The `current` property is defined as readonly, however it a valid way
+    ref.current = node
```

Comment:
May be someone knows React internals better. Is this valid thing?

https://github.com/facebook/react/blob/e58ecda9a2381735f2c326ee99a1ffa6486321ab/packages/react-reconciler/src/ReactFiberCommitWork.js#L618
https://github.com/facebook/react/blob/e58ecda9a2381735f2c326ee99a1ffa6486321ab/packages/react-reconciler/src/ReactFiberCommitWork.js#L629

I've checked React's source and they do the same thing in `commitAttachRef()` and `commitDetachRef()`

Reply:
https://reactjs.org/docs/hooks-reference.html#useref

> useRef returns a mutable ref object whose .current property is initialized to the passed argument (initialValue). The returned object will persist for the full lifetime of the component.

Reply:
yes, the later comment suggests that this is a safe move